src/controller/CHIPDeviceController.cpp

```diff
@@ -1178,6 +1184,298 @@ void DeviceCommissioner::OnSessionEstablished()
+    System::PacketBufferHandle attestationElementsTbs;
+
+    TLV::TLVReader tlvReader;
+    TLV::TLVType containerType = TLV::kTLVType_Structure;
+    int i                      = 0;
+
+    tlvReader.Init(attestationElements.data(), static_cast<uint32_t>(attestationElements.size()));
+    ReturnErrorOnFailure(tlvReader.Next(containerType, TLV::AnonymousTag));
+    ReturnErrorOnFailure(tlvReader.EnterContainer(containerType));
+    do
+    {
+        if (!TLV::IsContextTag(tlvReader.GetTag()))
+        {
+            continue;
+        }
+        const uint8_t * data = nullptr;
+        ReturnErrorOnFailure(tlvReader.GetDataPtr(data));
+        *element_array[i++] = ByteSpan(data, tlvReader.GetLength());
+    } while (tlvReader.Next() == CHIP_NO_ERROR);
+
+    // Step f
+    // Retrieve attestation challenge
+    PeerConnectionState * state = mSessionMgr->GetPeerConnectionState(
+        { mPairingSession.PeerConnection().GetPeerNodeId(), mPairingSession.GetPeerKeyId(), mAdminId });
+    VerifyOrReturnError(state != nullptr, CHIP_ERROR_NOT_CONNECTED);
+    ByteSpan attestationChallenge = state->GetSecureSession().GetAttestationChallenge();
+
+    // Create attestation-elements-tbs
+    attestationElementsTbs = System::PacketBufferHandle::New(System::PacketBuffer::kMaxSize);
+    VerifyOrReturnError(!attestationElementsTbs.IsNull(), CHIP_ERROR_NO_MEMORY);
+
+    tlvWriter.Init(std::move(attestationElementsTbs));
+    outerContainerType = TLV::kTLVType_NotSpecified;
+    ReturnErrorOnFailure(tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(0), attestationElements));
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(1), attestationChallenge));
+    ReturnErrorOnFailure(tlvWriter.EndContainer(outerContainerType));
+    ReturnErrorOnFailure(tlvWriter.Finalize(&attestationElementsTbs));
+
+    // Step e
+    P256ECDSASignature deviceSignature;
+    ReturnErrorOnFailure(deviceSignature.SetLength(signature.size()));
+    memcpy(deviceSignature, signature.data(), signature.size());
+    ReturnErrorOnFailure(mRemoteManufacturerPubkey.ECDSA_validate_msg_signature(
+        attestationElementsTbs->Start(), attestationElementsTbs->DataLength(), deviceSignature));
+
+    // Step g
+    VerifyOrReturnError(sizeof(mAttestationNonce) == attestationNonce.size(), CHIP_ERROR_INTERNAL);
+    VerifyOrReturnError(memcmp(mAttestationNonce, attestationNonce.data(), sizeof(mAttestationNonce)) == 0, CHIP_ERROR_INTERNAL);
+
+    // Step h/i
+    ReturnErrorOnFailure(ValidateCertificateDeclaration(certDeclaration, mRemoteManufacturerPubkey, firmwareInfo));
+
+    ChipLogProgress(Controller, "Sending Trusted Root Certificate to the device.");
+    ReturnErrorOnFailure(SendTrustedRootCertificate(device));
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DeviceCommissioner::ValidateCertificateDeclaration(chip::ByteSpan certDeclaration, Crypto::P256PublicKey pubkey,
```

Comment:
This is a Certifi**cation**Declaration", not a Certifi**cate**Declaration. Please rename.

Furthermore, the payloads are not signed specifically by the device's attestation private key. Therefore, the public key is not required.